### PR TITLE
feat(curation api): Enable 12 hour credentials.

### PR DESCRIPTION
**Pull request by Bento007**
### Reviewers
**Functional:** @danieljhegeman 

**Readability:** @metakuni <-- [Dan] Kuni we accidentally tagged Annie Ku instead 😂 

---

## Changes
- return upload path and bucket in POST /curator/v1/collections/{collection_uuid}/datasets/s3-upload-credentials
- add constant for super curator path
- add constant for super curator scope
- increase token expiration to 12 hours
- Add test function to generate an auth token without using Mock_Auth.

- connects to https://github.com/chanzuckerberg/single-cell-infra/pull/599
- connects to https://github.com/chanzuckerberg/single-cell-infra/pull/608
- zenhub issues https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/single-cell-data-portal/2508

## Known Issue
https://github.com/chanzuckerberg/single-cell-data-portal/issues/2554

### CODE REVIEW

File: backend/config/curation-api.yml [hunk 1]
```diff
@@ -408,6 +408,12 @@ paths:
       operationId: backend.corpora.lambdas.api.v1.curation.collection.dataset.post_s3_credentials
       parameters:
         - $ref: "#/components/parameters/path_collection_uuid"
+        - in: header
+          name: IdToken
+          schema:
+            type: string
+          required: true
+          description: Your id_token provided when generating an access token from
```
Comment (danieljhegeman) [inline backend/config/curation-api.yml:416]: fix

**Review by metakuni — APPROVED** [review]

File: backend/config/curation-api.yml [hunk 2]
```diff
@@ -408,6 +408,12 @@ paths:
       operationId: backend.corpora.lambdas.api.v1.curation.collection.dataset.post_s3_credentials
       parameters:
         - $ref: "#/components/parameters/path_collection_uuid"
+        - in: header
+          name: id_token
+          schema:
+            type: string
+          required: true
+          description: Your id_token provided when generating an access token from
```
[the remark this answers is not in the document]
Comment (danieljhegeman) [reply backend/config/curation-api.yml:416]: @Bento007

File: tests/unit/backend/corpora/api_server/curator/test_auth_token.py [hunk 3]
```diff
@@ -18,29 +18,28 @@ def test__post_token__201(self, auth0_management_session, CorporaAuthConfig):
         user_api_key = generate(test_user_id, test_secret)
         response = self.app.post("/curation/v1/auth/token", headers={"x-api-key": user_api_key})
         self.assertEqual(201, response.status_code)
-        token = response.json()["access_token"]
+        token = response.json["access_token"]
```
Comment (danieljhegeman) [inline tests/unit/backend/corpora/api_server/curator/test_auth_token.py:21]: intentional?

Comment (danieljhegeman) [reply tests/unit/backend/corpora/api_server/curator/test_auth_token.py:21]: I guess the tests are passing. I didn't think `json` was an attribute...

Comment (danieljhegeman) [reply tests/unit/backend/corpora/api_server/curator/test_auth_token.py:21]: Ohh... this isn't from the `requests` module. Nvm.

**Review by danieljhegeman — APPROVED** [review]

File: backend/config/curation-api.yml [hunk 4]
```diff
@@ -419,12 +419,16 @@ paths:
                   credentials:
                     type: object
                     properties:
-                      access_key:
```
Comment (Bento007) [inline backend/config/curation-api.yml:422]: fixing the response to match reality

File: backend/corpora/common/utils/authorization_checks.py [hunk 5]
```diff
@@ -10,11 +12,11 @@ def is_user_owner_or_allowed(user: str, scope: Union[list, str], owner: str) ->
     """
     Check if the user has ownership on a collection, or if it has superuser permissions
     """
-    return user == owner or (has_scope("write:collections", scope))
+    return user == owner or (has_scope(CorporaConstants.SUPER_CURATOR_SCOPE, scope))
```
Comment (metakuni) [inline backend/corpora/common/utils/authorization_checks.py:15]: Nice, I meant to suggest some constants a few PRs ago but forgot. 😄

File: backend/corpora/dataset_submissions/app.py [hunk 6]
```diff
@@ -46,6 +48,10 @@ def dataset_submissions_handler(s3_event: dict, unused_context) -> None:
             logger.debug(f"{collection_owner=}, {dataset_uuid=}")
             if not collection_owner:
                 raise CorporaException(f"Collection {collection_uuid} does not exist")
+            elif username != CorporaConstants.SUPER_CURATOR_NAME and username != collection_owner:
+                raise CorporaException(
+                    f"user:{username} does not have permission to modify datastes to {collection_uuid}."
```
Comment (metakuni) [inline backend/corpora/dataset_submissions/app.py:53]:
```suggestion
                    f"user:{username} does not have permission to modify datasets in collection {collection_uuid}."
```

**Review by metakuni — APPROVED** [review]

File: backend/config/curation-api.yml (hunk 4 above)
Comment (danieljhegeman) [reply backend/config/curation-api.yml:422]: do you want to add the upload path to the response in this PR?

File: tests/unit/backend/corpora/api_server/mock_auth.py [hunk 7]
```diff
@@ -144,6 +135,16 @@ def get_auth_token(app):
     return response.headers["Set-Cookie"]
 
 
+def make_token(token_claims: dict, token_duration: int = 0, additional_scope: list = None) -> dict:
+    addition_scope = additional_scope if additional_scope else []
```
Comment (danieljhegeman) [inline tests/unit/backend/corpora/api_server/mock_auth.py:139]: I'm not sure I like the rename to `addition_scope`. Confusing to me. Either reassign `additional_scope` or change to something visibly different.

File: tests/unit/backend/corpora/lambdas/dataset_submissions/test_dataset_submissions.py [hunk 8]
```diff
@@ -8,11 +8,16 @@
 
 class TestDatasetSubmissions(TestCase):
```
Comment (danieljhegeman) [inline tests/unit/backend/corpora/lambdas/dataset_submissions/test_dataset_submissions.py:9]: I think we should add `test__not_owner_of_collection__raises_error`

**Review by danieljhegeman — CHANGES_REQUESTED** [review]
A few comments but main issue is we need to add test to make sure the handler raises an error if the user doesn't own the collection.

File: backend/config/curation-api.yml (hunk 4 above)
Comment (Bento007) [reply backend/config/curation-api.yml:422]: It's already there.

File: tests/unit/backend/corpora/api_server/mock_auth.py (hunk 7 above)
Comment (Bento007) [reply tests/unit/backend/corpora/api_server/mock_auth.py:139]: I think what you're saying is you'd prefer the name to be scope?

Comment (Bento007) [reply tests/unit/backend/corpora/api_server/mock_auth.py:139]: actually, that's a typo.

File: backend/config/curation-api.yml (hunk 4 above)
Comment (danieljhegeman) [reply backend/config/curation-api.yml:422]: Oops 😝

File: tests/unit/backend/corpora/lambdas/dataset_submissions/test_dataset_submissions.py [hunk 9]
```diff
@@ -1,33 +1,68 @@
...
-
+        s3_event = create_s3_event(key=f"{self.user_name}/{self.collection_uuid}/{self.incoming_curator_tag}")
         mock_get_dataset_info.return_value = None, None
+        with self.assertRaises(CorporaException):
+            dataset_submissions_handler(s3_event, None)
 
+    @patch("backend.corpora.dataset_submissions.app.get_dataset_info")
+    def test__non_owner__raises_error(self, mock_get_dataset_info):
+        s3_event = create_s3_event(key=f"{self.user_name}/{self.collection_uuid}/{self.incoming_curator_tag}")
+        mock_get_dataset_info.return_value = "not_owner", self.dataset_uuid
```
Comment (danieljhegeman) [inline tests/unit/backend/corpora/lambdas/dataset_submissions/test_dataset_submissions.py:42]: 👍

File: tests/unit/backend/corpora/lambdas/dataset_submissions/test_dataset_submissions.py [hunk 10]
```diff
@@ -1,33 +1,68 @@
...
-        incoming_curator_tag = "my_dataset.h5ad"
-        s3_event = create_s3_event(bucket_name="expected_bucket", key=f"{collection_uuid}/{incoming_curator_tag}")
-
+        s3_event = create_s3_event(key=f"{self.user_name}/{self.collection_uuid}/{self.incoming_curator_tag}")
         mock_get_dataset_info.return_value = None, None
+        with self.assertRaises(CorporaException):
+            dataset_submissions_handler(s3_event, None)
 
+    @patch("backend.corpora.dataset_submissions.app.get_dataset_info")
+    def test__non_owner__raises_error(self, mock_get_dataset_info):
```
Comment (danieljhegeman) [inline tests/unit/backend/corpora/lambdas/dataset_submissions/test_dataset_submissions.py:40]: 👍

File: tests/unit/backend/corpora/api_server/mock_auth.py (hunk 7 above)
Comment (danieljhegeman) [reply tests/unit/backend/corpora/api_server/mock_auth.py]: cool, you'll fix?

**Review by danieljhegeman — APPROVED** [review]

**Pull request merged**